Extract JavaScript to validate from HTML files (fixes #764)

lahmatiy opened the pull request:
Implementation of `extract` option, that behave similar to jshint `extract` option (see #764 for details).
Solution tries avoid undesirable errors by `disallowTrailingWhitespace`, `disallowMixedSpacesAndTabs ` and `validateIndentation` rules.

aik099 commented:
:+1: Please also do the squash.

zxqfox commented:
I'm not sure but it's very close to inversion of jsx checking. And since that check uses custom esprima should do the extraction right in checker?

zxqfox commented:
btw. One more similar and common case is checking js inside md-files.

mrjoelkemp commented on lib/checker.js:
```diff
@@ -45,6 +46,38 @@ Checker.prototype.checkFile = function(path) {
+ * @returns {Promise * Errors}
+ */
+Checker.prototype.extractFile = function(path) {
+    if (!this._configuration.isFileExcluded(path)) {
```
```
if (!this._configuration.isFileExcluded(path)) {
  return Vow.resolve(null);
}

return vowFs.read(path, 'utf8').then(function(data) {
...
```

mrjoelkemp commented on lib/config/configuration.js:
```diff
@@ -148,6 +151,15 @@ Configuration.prototype.getFileExtensions = function() {
 /**
+ * Returns extract mode.
+ *
+ * @returns {String[]}
```
Seems to only return a `String` not a `String[]`.

mrjoelkemp commented on lib/extract-js.js:
```diff
@@ -0,0 +1,216 @@
+var Errors = require('./errors');
+
+/**
+ * Html file representation (needs for errors output).
```
`needed`?

mrjoelkemp commented on lib/extract-js.js:
```diff
@@ -0,0 +1,216 @@
+ *
+ * @name HtmlFile
+ */
+var HtmlFile = function(filename, source) {
```
Document params please

mrjoelkemp commented on lib/extract-js.js:
```diff
@@ -0,0 +1,216 @@
+     *
+     * @returns {String}
+     */
+    getFilename: function() {
```
Opt for `HtmlFile.prototype.getFilename`

lahmatiy replied:
You mean `HtmlFile.prototype.getFilename = function() { .. }` should be here?
I wrote this code as close as possible to [JsFile style](https://github.com/jscs-dev/node-jscs/blob/master/lib/js-file.js#L60).

mrjoelkemp commented on lib/extract-js.js:
```diff
@@ -0,0 +1,216 @@
+ * @returns {Object[]}
+ */
+function getScripts(html) {
+    function onopen(name, attrs) {
```
Consider extracting outside of `getScripts` with documentation of the function

mrjoelkemp replied:
Ah, I see that this depends on closed over vars later on. Tiny touch confusing, but no harm done. A class might be clearer, but not a blocker.

mrjoelkemp commented on lib/extract-js.js:
```diff
@@ -0,0 +1,216 @@
+function getScripts(html) {
+    function onopen(name, attrs) {
+        // tag should be a <script>
+        if (name !== 'script') {
```
Consider:

```js
// tag should be a <script>
if (name !== 'script' || 
    // ignore scripts with src attribute  
    attrs.src ||
    // script tag should has no type attribute or attribute should be equal to text/javascript
    (attrs.type && attrs.type.toLowerCase() !== 'text/javascript')) {
    return;
}
```

aik099 replied:
I'm not sure that we should even compare `type` attribute of `script` tag, because in HTML5 it defaults to `text/javascript` when missing.

mrjoelkemp commented on lib/extract-js.js:
```diff
@@ -0,0 +1,216 @@
+
+    function onclose() {
+        // if there is a script start pos
+        if (scriptStartPos) {
```
```
if (!scriptStartPos) {
    return;
}
```

mrjoelkemp commented on lib/extract-js.js:
```diff
@@ -0,0 +1,216 @@
+    }
+
+    function onclose() {
+        // if there is a script start pos
```
Unhelpful comment. Please remove

mrjoelkemp commented on lib/extract-js.js:
```diff
@@ -0,0 +1,216 @@
+}
+
+/**
+ * Trim script source and normalize lines offset.
```
@zxqfox Is there a jscs-jsdoc rule to enforce a newline after function description? Might be useful to add to #994

zxqfox replied:
Nope, but it's easy to make ;-) https://github.com/jscs-dev/jscs-jsdoc/issues/85

mrjoelkemp commented on lib/extract-js.js:
```diff
@@ -0,0 +1,216 @@
+    var line = 1;
+    var lastHtmlPos = 0;
+
+    for (var i = 0; i < scripts.length; i++) {
```
Opt for `scripts.forEach(function(scriptInfo) {...`

mrjoelkemp commented on lib/extract-js.js:
```diff
@@ -0,0 +1,216 @@
+ * by line offset. Result contains script sources with information about line
+ * offset (that was removed for each line) and lines count before script source.
+ * This information helps restore absolute positions in html file for errors.
+ * @param {String}
```
Finish the params documentation please.

mrjoelkemp commented on lib/extract-js.js:
```diff
@@ -0,0 +1,216 @@
+ * @param {String}
+ * @returns {Object[]}
+ */
+function extractJs(filename, data) {
```
Consider renaming `data` to `source` as it's more descriptive and matches the constructor's signature.

mrjoelkemp commented on lib/extract-js.js:
```diff
@@ -0,0 +1,216 @@
+ */
+var HtmlFile = function(filename, source) {
+    this._filename = filename;
+    this._lines = source.split(/\r\n|\r|\n/);
```
Can this regex be substituted with `/\r\n?|\n\r?/`? If so, extract this regex to a closed over var near the top of the file. It's used in 3 places.

zxqfox replied:
I think `/\r\n?|\n/` is enough. What is `\n\r`?

aik099 replied:
See there is `|` in there, so we're catching any wild `\n` or `\r`.

zxqfox replied:
@aik099 Line end is a `\r` (mac) or `\n` (unix) or `\r\n` (win). We can tune for gracefully parse mixes of them, but it really can be substituted with `/\r\n?|\n/`. Right?

aik099 replied:
`\r` isn't used on Mac since ages :wink: To deal with files, where mixed line endings exists I usually do this:

1. replace `\r\n` with `\n`
2. replace `\r` with `\n`
3. split by `\n`

mrjoelkemp commented on lib/extract-js.js:
```diff
@@ -0,0 +1,216 @@
+        // add line offset before script
+        line += data.substring(lastHtmlPos, scriptInfo.start).split(/\r\n?|\n\r?/).length - 1;
+
+        // store result
```
Unnecessary comment. Please remove

mrjoelkemp commented on lib/extract-js.js:
```diff
@@ -0,0 +1,216 @@
+        sources: sources,
+        errors: errors,
+        addError: function(error) {
+            errors._errorList.push({
```
`errors.add`?

mrjoelkemp replied:
Consider:

```js
addError: function(error) {
    error.filename = filename;
    errors.add(error);
}
```

lahmatiy replied:
It doesn't work as you expected https://github.com/jscs-dev/node-jscs/blob/master/lib/errors.js#L34
I'll be happy if it does

markelog replied:
Hm, it seems we need to create a public push method

mrjoelkemp commented on lib/extract-js.js:
```diff
@@ -0,0 +1,216 @@
+    var offset;
+
+    // remove first list if it's an empty string
+    // usualy <script> starts with new line
```
`usually`

mrjoelkemp commented on lib/extract-js.js:
```diff
@@ -0,0 +1,216 @@
+            var source = html.substring(scriptStartPos, scriptEndPos);
+
+            // store script content only if it contains non-whitespace characters
+            if (/\S/.test(source)) {
```
Consider extracting `/\S/` to a var for reuse. The var name should explain what the regex represents. The comment above could likely be removed in that case.

mrjoelkemp commented on lib/extract-js.js:
```diff
@@ -0,0 +1,216 @@
+
+    // replace last line by empty string if it contains only whitespaces
+    // it helps avoid disallowTrailingWhitespace errors on last line
+    lines.push(/\S/.test(lastLine) ? lastLine : '');
```
Why pop the last line only to push it back on? Why not just do something like:

```js
var lastLine = lines[lines.length - 1];
...
if (!/\S/.test(lastLine)) {
  lines[lines.length - 1] = '';
}
```

mrjoelkemp commented on lib/extract-js.js:
```diff
@@ -0,0 +1,216 @@
+            return Infinity;
+        }
+
+        // fetch whitespaces at the line begining
```
`beginning`

mrjoelkemp commented on lib/extract-js.js:
```diff
@@ -0,0 +1,216 @@
+ * @param {String} source
+ * @returns {Object[]}
+ */
+function normalizeSource(source) {
```
I'm not understanding why this function is necessary. Mind expanding on this function's docblock explaining why?

mrjoelkemp commented on lib/checker.js:
```diff
@@ -60,15 +93,22 @@ Checker.prototype.checkDirectory = function(path) {
+                    if (this._configuration.getExtract() === 'never') {
+                        return [];
+                    }
+                    promise = Vow.when(this.extractFile(fullname));
```
Wouldn't you need to check that `fullname` contains the `.html` extension? Is that check done somewhere else? Seems like this would attempt to parse *everything* but `.js` files.

lahmatiy replied:
I'm not sure about this. jshint does the same as I remember

markelog replied:
Wouldn't that slows us down? If user does something like `jscs .`? Then read we would read every perky out there.

markelog replied:
How about extending defaults of `fileExtensions` option instead of introducing the `extract` and putting every possible html extension?

It has obvious downside, like if user redefine it, their would need to add those extension again,

lahmatiy replied:
We could use `never` as default value for `extract` option as `JSHint` has. I believe we should do initial solution as simple as possible. Additional options (and other improvement) could be added later.

If somebody runs `jscs .` it's not a problem I guess. `never` for `extract` by default should does no any performance impact.

markelog replied:
If `JSHint` did it, it doesn't mean it is a good solution, with `extract = never` by default, user would need to know about this option, with `extract = auto` by default, it would hurt performance.

Whereas if we would extend `fileExtensions` we would support most use-cases performance downside.

> I believe we should do initial solution as simple as possible.

If we would add `extract` option we need to support it for long time.

lahmatiy replied:
Maybe add support for array of minimatch masks in additional of `never`, `auto` and `always`? And set `['*.html', '*.htm', '*.xhtml']` or `['*.(html?|xhtml)']` by default?

```
{
  "extract": ['*.html', '*.htm', '*.xhtml']
}
```

```
jscs . --extract *.html *.xhtml
```

markelog replied:
Sounds good!

mrjoelkemp commented on lib/checker.js:
```diff
@@ -45,6 +46,38 @@ Checker.prototype.checkFile = function(path) {
+Checker.prototype.extractFile = function(path) {
+    if (!this._configuration.isFileExcluded(path)) {
+        return vowFs.read(path, 'utf8').then(function(data) {
+            if (this._configuration.getExtract() === 'always' || /^\s*</.test(data)) {
```
Comment on what the regex is or extract to a var with a name that explains it.

mrjoelkemp commented on lib/checker.js:
```diff
@@ -45,6 +46,38 @@ Checker.prototype.checkFile = function(path) {
+ */
+Checker.prototype.extractFile = function(path) {
+    if (!this._configuration.isFileExcluded(path)) {
+        return vowFs.read(path, 'utf8').then(function(data) {
```
Consider renaming `data` to `source` for clarity

mrjoelkemp commented on lib/checker.js:
```diff
@@ -45,6 +46,38 @@ Checker.prototype.checkFile = function(path) {
+                    });
+                }, this);
+
+                if (!content.errors.isEmpty()) {
```
Need to return something for the case when it's empty? Would `null` suffice?

mrjoelkemp commented on lib/checker.js:
```diff
@@ -45,6 +46,38 @@ Checker.prototype.checkFile = function(path) {
+ * Extract JavaScript from file.
+ *
+ * @param {String} path
+ * @returns {Promise * Errors}
```
I believe we typically do `@returns {Promise} descriptionHere` where the description can describe the resolved values. In this case it's an error list or null.

mrjoelkemp commented on test/checker.js:
```diff
@@ -47,6 +47,14 @@ describe('modules/checker', function() {
         });
+
+        it('should throw an exception when path doesn\'t exists', function() {
+            return checker.checkPath('./test/non-exists').then(function() {
```
An alternative might be:

```js
return checker.checkPath('./test/non-exists').then(null, done);
```

This requires `done` as a parameter to the `it`, of course

zxqfox replied:
I'm not sure but looks like mocha has built-in promise support.

mrjoelkemp commented on test/checker.js:
```diff
@@ -78,4 +86,107 @@ describe('modules/checker', function() {
+            });
+
+            return checker.extractFile('./test/data/extract/always.html').then(function(errors) {
+                assert(errors === null);
```
Why return `null` instead of an empty list?

mrjoelkemp commented on test/extract-js.js:
```diff
@@ -0,0 +1,71 @@
+var assert = require('assert');
+var Checker = require('../lib/checker');
+
+describe('modules/extract', function() {
```
`modules/extract-js` although `lib/extract-js` would be more accurate (though, I'm aware we have the discrepancy for other test suites).

mrjoelkemp commented on test/extract-js.js:
```diff
@@ -0,0 +1,71 @@
+                    errorsCollection.forEach(function(errors) {
+                        if (!errors.isEmpty()) {
+                            errors.getErrorList().forEach(function(error) {
+                                errors.explainError(error, true);
```
Does this simply test that the error format added within extract-js is correct?

lahmatiy replied:
I had an issue when reporter crashes for somehow. So I'd add this test.
But I don't remember details. Probably this could be removed...

lahmatiy replied:
Ah... I just remembered! Reporter crashes as we use custom file class (`HtmlFile`) in errors. So error details should be compatible with console reporter (other reporters ok, but console reporter not, because do additional formatting). Something like that.

mrjoelkemp commented on test/extract-js.js:
```diff
@@ -0,0 +1,71 @@
+            });
+            return checker.extractFile('./test/data/extract/index.html').then(function(errors) {
+                // disallowMixedSpacesAndTabs doesn't warn when some line contains only and some
+                // only spaces (i.e. '\s\svar foo;\t\tvar bar;'). If this behavior will change
```
Sounds like an issue should be filed.

mrjoelkemp replied:
Would you mind adding an issue for this with what you expect to see from the rule?

zxqfox replied:
:+1:

lahmatiy replied:
Well, I expect that if some lines has tabs offset but others spaces offset – warning should be shown. For now nothing happen. So, here is we will see 2 warnings:

```
\s\t
\t\s
```

But in this case no warnings:

```
\s\s
\t\t
```

But probably it's a feature, not a bug ;)

mrjoelkemp commented:
@lahmatiy This is really impressive work! Great job and thanks for contributing!

Please rebase off of master. 

I added a bunch of comments, most of which are suggestions to increase the clarity of the code. Feel free to ask questions about any of the comments.

After the comments are addressed, we'll need another maintainer to review the code before merging. It's a big PR, so we need more than one review on this one.

You're also missing changes to Overview.md: https://github.com/jscs-dev/node-jscs/blob/master/OVERVIEW.md#options to document the `extract` option and its possible values.

I noticed (and you stated) that the algorithm is similar to jshint's approach but the code is original, so I think we're okay with licensing.

Awesome work!

lahmatiy commented:
@mrjoelkemp Thank you. I'll fix PR asap.

I read your comments. And I would like to notice that I tried write the code as close as possible to original codebase. I prefer code consistency even when I see that some code could be written better. Otherwise other similar code fragments should be changed too, but it's out of bounds of this PR.

I also worrying about default value for `extract`. JSHint has `auto` by default. But the same setting for `jscs` might break build of projects that using `jscs`, as `jscs` will start checking something that didn't check before. Probably `never` should be better for backward capability. In this case `extract` will be opt-in.

aik099 commented:
Maybe you should also rebase, because GitHub showing that PR can't be merged due conflicts.

lahmatiy commented:
@aik099 jscs codebase changes too fast, that PRs becomes broken all the time :( Sure I'll rebase, when return to work on feature.

@mdevils I believe your should take a look on this PR. As I know you're refactoring `JsFile` class. For this feature we need some common `File` abstraction. Because we need to parse fragments of some file and it isn't a JavaScript file. I've made a simple mock class `HtmlFile` for this purposes (temporary solution). But I'm sure there should be better solution.

zxqfox commented:
@lahmatiy Better solution is to use custom esprima imo.

aik099 commented:
> jscs codebase changes too fast

That is something each open source project can wish for. As long as not BC breaks are introduced.

lahmatiy commented:
@zxqfox How it could help to parse `html`, find out `js` fragments and aggregate fragment errors in one error list mapped on original file?

zxqfox commented:
@lahmatiy It can skip all html nodes as comments (or anything concrete) and returns just Parser API AST nodes to for checking. But I agree that it will be a problem with indentation probably.

lahmatiy commented:
@zxqfox It's not a working solution. We need text of correct `<script>` tags only. And we should know where fragment starts and ends. It requires for normalization, as rules check indentation, empty lines, trailing whitespaces, trailing new line, `var` position and count etc.

mikesherov commented:
> @zxqfox It's not a working solution. We need text of correct <script> tags only. And we should know where fragment starts and ends. It requires for normalization, as rules check indentation, empty lines, trailing whitespaces, trailing new line, var position and count etc.

@lahmatiy the correct way to handle this then seems to make the rules that check this stuff aware of that fact. For example, the indentation problem is easily fixed by making the rule not assume that 0 indents mean column 1. It can easily check the indentation of the first token it encounters and set that as the offset of the code. 

For the rest of the issues, it seems like a possible solution is remove all non script tags, replace it all with whitespace, and add `/*jscs:disable*/` to the top of the document, replace `<script>` with `/*jscs:enable*/`, and `</script>` with `/*jscs:disable*/`.
 
For example:

```
<html>
....
<body>
yada
  <script>
  var a=1;
  </script>
  dada
  <script>
  var b=2;
  </script>
```

becomes:
```
/*jscs:disable*/



  /*jscs:enable*/
  var a=1;
  /*jscs:disable*/

  /*jscs:enable*/
  var b=2;
  /*jscs:disable*/
```

this will automatically ignore errors anywhere that isn't actual scripts tags, and automatically retain position information. All that needs to be adjusted is again, the offset for indentation.

zxqfox commented:
@lahmatiy Em. Why not?
For simple html:
```

for 
you'll have AST like that:
```js
comments (or something):
[
  { type: 'HTMLPart', value: '<html>...<script></script><script>\n</script>...<script>', range: ... },
  { type: 'HTMLPart', value: '</script>etc...</html>' }
]
nodes:
[
  { type: 'SomeStandardNode', value: 'var', etc... }, ...
]
```

The only thing to fix here is lines/columns normalization for indentation and reporters. For errors

lahmatiy commented:
@mikesherov @zxqfox
I'm not sure that custom parser could strip markup correctly. There are many cases for wrong matches if you don't process source as html, but js with possible markup inside.
We can't merge all scripts in single one. We should process scripts independantly (as separate files, as browsers do). Otherwise one broken script may broke others, or fix it, or change sematic of next script, for example:
```html
<script>
  (function(){ /* I'm IIFE */ })()
</script>
<script>
  (function(){ /* I'm IIFE */ })()
</script>
```

Indentations may vary from `<script>` to `<script>`, so fixing `validateIndentation` rule couldn't help here.
And there are some rules that could find an errors, because of script merge and no normalization:
- disallowPaddingNewlinesBeforeKeywords
- disallowSemicolons
- disallowTrailingWhitespace
- requireMultipleVarDecl
- requireParenthesesAroundIIFE
- requireSemicolons (PR)
- validateIndentation
- validateQuoteMarks: true

etc.

mikesherov commented:
@lahmatiy those are good points. Still would like to see this code somehow simplified. Adds a lot of complexity at the moment.

mdevils commented:
@lahmatiy I will take a look soon.

mdevils commented:
We discussed this with @lahmatiy, this implementation looks like a hack right now. We need better abstraction for this case (which will also support autofixing).

I'm going to work on this later, but it might require to make a `2.0` as API rebuild might be necessary.

What do you think?

zxqfox commented:
:+1: Go ahead!

lahmatiy commented:
@zxqfox @mdevils Checking code inside jsdocs in @example section would be useful too ;)

markelog commented:
How about we settle on this version right now and rewrite it if needed?

zxqfox commented:
@markelog It would be great to have an abstraction for these purposes...

markelog commented:
There is stuff to improve, but i would rather do it later, get response from the community and wait three months

nicolasgarnier commented:
Any news on this feature? We want to add JSCS checks as part of the Polymer Starter Kit's build pipeline so we're trying to assess if we should go for a workaround or wait for this.

Cheers!

lahmatiy commented:
@mdevils any news about new abstractions for code fragments? Maybe we could merge this feature for html only with future improvements (js fragments in any content type) when ready?
/cc @mikesherov @markelog

markelog commented:
We should prioritize this and schedule for the new release

lahmatiy commented:
Ok, I will rebase pr and fix review notes then.
When do you plan next release?

markelog commented:
Whenever it's ready :-), it seems it could be soon though, there is couple prs we could land

nicolasgarnier commented:
Awesome thanks guys. We'll keep an eye on this PR and hope for the best :)

lahmatiy commented:
Mostly done.
If everything is ok I'm squash commits then.

markelog commented on lib/extract-js.js:
```diff
@@ -0,0 +1,231 @@
+            // ignore scripts with src attribute
+            attrs.src ||
+            // script tag should has no type attribute or attribute should be equal to text/javascript
+            (attrs.type && attrs.type.toLowerCase() !== 'text/javascript')) {
```
`text/javascript`  is obsolete, check this out - http://tools.ietf.org/html/rfc4329

lahmatiy replied:
I know, but still valid and may be used in sources.
This condition checks that `<script>` has no `type` other than `text/javascript`.

```html
<script></script>  <!-- process -->
<script type="text/javascript"></script>  <!-- process -->
<script type="some/template"></script>    <!-- ignore -->
```

markelog replied:
Yes, sure, but i provided the link not for nothing, what about - 

```
application/ecmascript
application/javascript
application/x-ecmascript
application/x-javascript
text/ecmascript
text/javascript
text/javascript1.0
text/javascript1.1
text/javascript1.2
text/javascript1.3
text/javascript1.4
text/javascript1.5
```

lahmatiy replied:
I suppose values other than `text/javascript` is really rare used. If you insist, I can add check for all of those values. But I thought it's good idea to start with `text/javascript` and add new values if needed depending on users feedback.

markelog replied:
Would be good, but yeah, we could do it later

markelog commented on test/specs/checker.js:
```diff
@@ -80,6 +88,109 @@ describe('checker', function() {
+                extract: 'never'
+            });
+
+            return checker.checkPath('./test/data/extract').then(function(errors) {
```
What if pass a full path?

lahmatiy replied:
Should I add `checkPath` and `checkFile` versions for every single test?

markelog replied:
I was asking would happen if we would pass a full path with extension

markelog commented:
`extract`
So this is not a cli flag? What about documentation? Do we really needed? JSHint has it, but i wonder if it is redundant?

markelog commented on lib/checker.js:
```diff
@@ -1,6 +1,7 @@
 var vowFs = require('vow-fs');
 var Vow = require('vow');
 var StringChecker = require('./string-checker');
+var extractJs = require('./extract-js');
```
So this class called as `HtmlFile` analogues to `JsFile`, but file named as `extract-js`?

lahmatiy replied:
`HtmlFile` is just a helper for `extract` option, because we can't use `JsFile`, but errors must has a reference to some `file`.

markelog replied:
Fair enough

lahmatiy commented:
@markelog 

> So this is not a cli flag? What about documentation?

No requests before, but I could add both if needed. But we need to determine your interest first. PR is 7 months old already, but opinion about feature is still not clear to me.

> Do we really needed? JSHint has it, but i wonder if it is redundant?

Yes, it's needed. I saw feature requests before. And one day I needed this feature too. That's why I implemented it and thought it might be helpful for others too. You can see some attached issues here. I don't understand why you think it might be redundant?

Most strange part to me, that you ask those questions after your own proposition recently to continue work on it (and proposition about merge in june) and when I fixed known problems once again :-/

markelog commented:
> No requests before, but I could add both if needed. But we need to determine your interest first. PR is 7 months old already, but owners opinion about feature is still not clear to me.

No-no, that was about replacing it with `fileExtensions` property.

> PR is 7 months old already, but owners opinion about feature is still not clear to me.

I would landed it it seven months ago, but couldn't, since some of us thought we should wait, in other words, sometimes, perfectionism wins when it shouldn't :-(.

> Most strange part to me, that you ask those questions after your own proposition recently to continue work on it (and proposition about merge in june) and when I fixed known problems once again :-/

Sorry for misunderstanding, i should have better express my thoughts, i'm very excited about this feature,  one way or another, it will be in the next release.

lahmatiy commented:
Oh, I see. Thank you for your clarifications.

I implementing cli option and array of masks support then. But I'm not sure about documentation, should I add it here https://github.com/jscs-dev/jscs-dev.github.io/blob/master/overview.html?

markelog commented:
[Here](https://github.com/jscs-dev/node-jscs/blob/master/OVERVIEW.md) would be good

lahmatiy commented:
OK. Will be done.

markelog commented:
Would like to release new version this week, will you have time to finish this up?

lahmatiy commented:
I have some problems with merge now. Even after merge conflict resolving, some tests are failed because of changes in master :( I tried to fix it today but no success yet.
I afraid I have no time till Sunday. But on Sunday I would continue to work on it (maybe earlier but I'm not sure).

markelog commented:
> I afraid I have no time till Sunday. But on Sunday I would continue to work on it (maybe earlier but I'm not sure).

Friendly ping

lahmatiy commented:
Sorry for delay.
Planning to continue work later today.

lahmatiy commented:
Looks done.
Awaiting for review.

markelog commented on OVERVIEW.md:
````diff
@@ -148,6 +148,25 @@ jscs path[ path[...]] --preset=jquery
+jscs path[ path[...]] --extract *.html --extract *.htm
+```
+
+Or you could you one of shortcut value: `always`, `auto` or `never`. Note that this values couldn't be comined with file masks.
````
`comined`

markelog commented on OVERVIEW.md:
````diff
@@ -148,6 +148,25 @@ jscs path[ path[...]] --preset=jquery
+jscs path[ path[...]] --extract always
+```
+
+Currently only `html` format is supported (JavaScript inside `<script>` will be checked). Auto fixing is not apply.
````
` Auto fixing is not apply` - > `Auto fixing does not apply`?

markelog replied:
`Currently only `html` format...` what else is there? `xml`? or...?

lahmatiy replied:
We discussed about extracting js from markdown or jsdocs, as example.

markelog commented on OVERVIEW.md:
```diff
@@ -281,6 +300,30 @@ Values: A single file extension or an Array of file extensions, beginning with a
+
+Type: `Array` or `"always"` or `"auto"` or `"never"`
+
+JavaScript extracting from files that doesn't match to `fileExtensions` or `excludeFiles`. Only `html` format supported.
```
Maybe other options could be links?

markelog commented on OVERVIEW.md:
```diff
@@ -281,6 +300,30 @@ Values: A single file extension or an Array of file extensions, beginning with a
+
+JavaScript extracting from files that doesn't match to `fileExtensions` or `excludeFiles`. Only `html` format supported.
+
+Special values meaning:
```
Would remove `meaning`

lahmatiy commented:
For now I'm not sure we need for special values, i.e. `always`, `auto` and `never`. It looks odd. Maybe I should remove it?

markelog commented on OVERVIEW.md:
```diff
@@ -281,6 +300,30 @@ Values: A single file extension or an Array of file extensions, beginning with a
+JavaScript extracting from files that doesn't match to `fileExtensions` or `excludeFiles`. Only `html` format supported.
+
+Special values meaning:
+- `always` - JavaScript extracts from any file
```
`JavaScript extracts from any file` -> `extract from any file`, since you use imperative mood down the road

markelog commented on OVERVIEW.md:
```diff
@@ -281,6 +300,30 @@ Values: A single file extension or an Array of file extensions, beginning with a
+
+#### Default
+
+Default value is `["**/*.+(htm|html|xhtml)"]`. So JavaScript extracts from files with `.htm`, `.html` or `.xhtml` extension.
```
`So JavaScript extracts` -> `So JavaScript is extracted`?

markelog commented on bin/jscs:
```diff
@@ -36,3 +38,25 @@ program
+ * Collect array of file mask. It also allow single special value
+ * for this option.
+ */
+function extract(value, memo) {
```
It would be better to not have any logic in bin file, would you mind moving it?

markelog commented on lib/checker.js:
```diff
@@ -116,12 +151,19 @@ Checker.prototype._processDirectory = function(path, fileHandler) {
+                    }
+
+                    return this.extractFile(fullname).then(function(errors) {
+                        // extractFile could don't extract js from file
```
`could don't`?

markelog commented on lib/checker.js:
```diff
@@ -147,14 +188,14 @@ Checker.prototype._processPath = function(path, fileHandler) {
         return vowFs.stat(path).then(function(stat) {
             if (stat.isDirectory()) {
-                return _this._processDirectory(path, fileHandler);
+                return this._processDirectory(path, fileHandler);
```
You really like `this`, huh? :-)

lahmatiy replied:
Just align to style in other methods.

markelog commented:
> For now I'm not sure we need for special values, i.e. always, auto and never.

:thumbsup:

markelog commented:
Would be cool if anyone else would check docs, /cc @hzoo, @mikesherov

hzoo commented on OVERVIEW.md:
```diff
@@ -148,6 +148,25 @@ jscs path[ path[...]] --preset=jquery
 In order to add/remove preset rules you will need to create a `.jscsrc` config file.
 
+### `--extract`
+With this option you could set file mask from which JavaScript should be extracted.
```
` you could set file ` -> ` you can set the file `

hzoo commented on OVERVIEW.md:
````diff
@@ -148,6 +148,25 @@ jscs path[ path[...]] --preset=jquery
+jscs path[ path[...]] --extract *.html 
+```
+
+You could set several masks if needed.
````
`could set` -> `can set` 
`if needed` -> `if necessary`

lahmatiy commented:
Ok, I'm going to remove always/auto/never. It's simplify implementation.

hzoo commented on OVERVIEW.md:
````diff
@@ -148,6 +148,25 @@ jscs path[ path[...]] --preset=jquery
+jscs path[ path[...]] --extract *.html --extract *.htm
+```
+
+Or you could you one of shortcut value: `always`, `auto` or `never`. Note that those values couldn't be combined with file masks.
````
`you could you one of` -> `you can use a`.

`Shortcut values cannot be combined with file masks`

hzoo commented on OVERVIEW.md:
````diff
@@ -148,6 +148,25 @@ jscs path[ path[...]] --preset=jquery
+jscs path[ path[...]] --extract always
+```
+
+Currently only `html` format is supported (JavaScript inside `<script>` will be checked). Auto fixing does not apply.
````
Currently, only the `html` format is supported (JavaScript inside `<script>` will be checked) and extracted code cannot be auto fixed.

hzoo commented:
Yeah I'l look at the docs.

I guess I'm not familiar with the term - is it normally `file masks`? Does file extension work as well?

hzoo commented on OVERVIEW.md:
```diff
@@ -148,6 +148,25 @@ jscs path[ path[...]] --preset=jquery
 In order to add/remove preset rules you will need to create a `.jscsrc` config file.
 
+### `--extract`
+With this option you can set file mask from which JavaScript should be extracted.
```
Sorry I updated my comment - `set file mask` -> `set the file mask`. Actually the use of `from which` should probably be changed. Maybe `This option checks javascript style in the specified file extensions (currently only html)` or something like that.

lahmatiy replied:
I don't how to describe it right :-/
It's not about file extensions, but glob patterns, I guess.

lahmatiy commented:
Well, I may be wrong about `file masks`. Actually I mean `glob patterns` as I can see in `excludeFiles`.

lahmatiy commented:
Done! No more always/auto/never :)

lahmatiy commented:
Feature looks completed.
Should I do anything else?
 /cc @markelog @mikesherov

markelog commented:
Okay, i see there is couple artefacts of the previous logic and room for improvement, but we can deal with that later if that would be an issue

markelog commented:
Would you mind squashing your commits?

lahmatiy commented:
Done

markelog commented:
Hooray! 

Many thanks @lahmatiy, this was truly epic patience test!

hzoo commented:
Truly :tada:

lahmatiy commented:
\o/ it's finally merged :)
